python_coreml_stable_diffusion/activation_quantization.py

```diff
@@ -0,0 +1,489 @@
+
+    config = LinearQuantizerConfig(
+        global_config=ModuleLinearQuantizerConfig(
+            quantization_scheme="symmetric",
+            milestones=[0, 1000, 1000, 0],
+        ),
+        module_name_configs=module_name_config,
+        module_type_configs={
+            torch.cat: None,
+            torch.nn.GroupNorm: None,
+            torch.nn.SiLU: None,
+            torch.nn.functional.gelu: None,
+            operator.add: None,
+        },
+    )
+    return config
+
+def quantize(model, config, calibration_data):
+    """
+    Apply post training activation quantization to specified model, using calibration data
+    """
+    submodules = dict(model.named_modules(remove_duplicate=True))
+    layer_norm_modules = [key for key, val in submodules.items() if isinstance(val, LayerNormANE)]
+    non_traceable_module_names = layer_norm_modules + [
+        "time_proj",
+        "time_embedding",
+    ]
+
+    # Mark certain modules as non-traceable to make the UNet model fx traceable
+    config.non_traceable_module_names = non_traceable_module_names
+    config.preserved_attributes = ['config', 'device']
+
+    sample_input = calibration_data[0]
+    quantizer = LinearQuantizer(model, config)
+    logger.info("Preparing model for quantization")
+    prepared_model = quantizer.prepare(example_inputs=(sample_input,))
+    prepared_model.eval()
+
+    quantizer.step()
+
+    logger.info("Calibrate")
+    for idx, data in enumerate(calibration_data):
+        logger.info(f"Calibration data sample: {idx}")
+        prepared_model(*data)
+
+    logger.info("Finalize model")
+    quantized_model = quantizer.finalize()
+    return quantized_model
+
+def get_quantizable_modules(unet):
+    quantizable_modules = []
+    for name, module in unet.named_modules():
+        if len(list(module.children())) > 0:
+            continue
+        if type(module) == torch.nn.modules.conv.Conv2d:
+            quantizable_modules.append(('conv', name))
+        if type(module) == Einsum:
+            quantizable_modules.append(('einsum', name))
+
+    return quantizable_modules
+
+def register_input_log_hook(unet, inputs):
+    """
+    Register forward pre hook to save model inputs 
+    """
+    def hook(_, input):
+        input_copy = deepcopy(input)
+        input_copy = tuple(i.to('cpu') for i in input_copy)
+        inputs.append(input_copy)
+
+        # Return inputs unmodified
+        return input
+
+    return unet.register_forward_pre_hook(hook)
+
+def generate_calibration_data(pipe, args, calibration_dir):
+    # Register forward pre hook to record unet inputs
+    unet_inputs = []
+    handle = register_input_log_hook(pipe.unet, unet_inputs)
+
+    # If directory doesn't exist, create it
+    os.makedirs(calibration_dir, exist_ok=True)
+
+    # Run calibration prompts through the pipeline and
+    # serialize recorded UNet model inputs
+    for prompt in CALIBRATION_DATA:
+        gen = torch.manual_seed(args.seed)
```

Comment:
any reason to set the seed inside the for loop, instead of say, globally once?

Reply:
Just to keep the output of the pipeline for any prompt to be associated with a specific seed. For example, the output data for a given prompt would not vary based on how many prompts are used.